### ISSUE TITLE
git-lfs handling of gitignore for directories leads to git operation slowdowns

### ISSUE THREAD

**Issue by whitty** [git-lfs/git-lfs#2221]
I'm tracking down performance degradations that match some observations from #1750.

We have `.gitignore` entries at the base for output directories that can exist at any stage of the tree - ie `obj/...`, `a/b/obj/...` `c/obj/...` , which are ignored via `.gitignore` rule at the base of the tree

```
# Output directories
obj/
```

Running

```
$ strace -f -o log git lfs track
```

Shows that the obj-directories are being visited (hint some of these directories have absurdly big contents - think full linux kernel tree)

From https://github.com/git-lfs/git-lfs/issues/1750#issuecomment-267281394 the suggestion is to use `**/obj/**` which does indeed properly exclude these directories from git-lfs thus dropping track time from 9 seconds to 0.048 seconds.  (note 9 seconds is more like 30 seconds when the disk-cache isn't warm).  That 9 second duration is incurred on every step of a rebase.

Now the comment about using `**/dir/**` was made "in-passing", but it seems having different interpretation of the `.gitignore` file compared to git is a bug of some sort, that either needs:

1. Justification that git its not an error
  1. and thus potentially a programmatic warning in git-lfs 
  1. but definitely documentation or best practices
2. Deferring
  1. and again documentation
3. Fixing

I couldn't find anywhere that this issue was tracked, hence I've logged this.

```
$ git version
git version 1.9.1

$ git lfs version
git-lfs/2.1.0 (GitHub; linux amd64; go 1.8.1)

$ lsb_release -a
...
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.5 LTS
Release:	14.04
Codename:	trusty
```

**Comment by technoweenie** [comment]
Thanks for pointing this out. I think the problem here is that `.gitignore` and `.gitattributes` use the same syntax, but interpret it differently. When given `obj/`, `.gitignore` will count any `obj/**/*` file as a match and ignore it. However, `.gitattributes` requires a more exact match. I think [this](https://github.com/git-lfs/git-lfs/blob/6400df65b69a564f1982f608311a6d9520ba727d/tools/filetools.go#L282-L289) would be the proper place for a fix.

**Comment by technoweenie** [comment]
Hey @whitty: the fix is in v2.1.1.  How's it perform now? Did it help?

**Comment by whitty** [comment]
Brilliant - strace says that obj directories were not traversed - previously the **/obj/** workaround rule had still at least enumerated (`lstat`) the first level below obj, now only `obj` itself is statted.

(note -after encountering this bug did heavy workspace cleanup so my number aren't as dramatic as the original 9 second case)

One workspace showed `time` results for `git lfs track` dropping from

```
real	0m1.637s
user	0m11.476s
sys	0m0.356s
```

to

```
real	0m0.032s
user	0m0.048s
sys	0m0.024s
```

and those numbers are after having filled the fs-cache.  The first run on a cold workspace where before this change had `real	0m37.936s` (yes we use spinning disks still).

And a further comparison - 2.1.0 with **/obj/** workaround

```
real	0m0.074s
user	0m0.200s
sys	0m0.008s
```

vs 2.1.1

```
real	0m0.021s
user	0m0.048s
sys	0m0.008s
```

so processing (user time) is dropped by an order of magnitude.

Beautiful :)
